### ISSUE TITLE
Hot reload in dev-server

### ISSUE THREAD

**Issue by thangngoc89** [MoOx/statinamic#140]
With the current fix in `next` branch https://github.com/MoOx/statinamic/commit/3fabb759363fa536ac518a1b95eadb291733ebef user will receive a **false** complaint about invalid checksum.

Why ?

- Client receive the lastest (hot loaded) script ?
- Dev server doesn't

**Comment by MoOx** [comment]
Dev server cannot "receive" a hot update. It's running using node, not webpack (even if a require hook make some transformation independently).
The easiest thing to is to clean some pieces of the node cache when "some files" are update.
I used this technique with metalsmith-watch for an accurate rerender with metalsmith-react.

Also please consider this note https://github.com/FWeinb/metalsmith-watch/blob/4bd763f781466e024b4c3942f43c77d13182ad31/src/index.es#L22-L24
https://github.com/FWeinb/metalsmith-watch/blob/4bd763f781466e024b4c3942f43c77d13182ad31/src/index.es#L255-L278

Tests https://github.com/FWeinb/metalsmith-watch/blob/4bd763f781466e024b4c3942f43c77d13182ad31/__tests__/cache.es

Do you think that can do the thing?

**Comment by thangngoc89** [comment]
I'll give it a try. Not so sure

**Comment by thangngoc89** [comment]
I'm think of using webpack as file watcher. Reading webpack plugin docs right now

**Comment by MoOx** [comment]
Why do you want to watch files? We should just (stupidly) clean node cache before generating a page in dev server, no need to watch anything imo.

**Comment by thangngoc89** [comment]
I'm doing it. But have no idea where to put this ?
```js
// Invalidate all files
  Object.keys(require.cache)
    .filter((t) =>
      t.startsWith(config.cwd) &&
      !t.startsWith(join(config.cwd, 'node_modules'))
    )
    .forEach((t) => {
      console.log(t)
      delete require.cache[t]
    })
```

**Comment by thangngoc89** [comment]
I give up. Can't make it work

**Comment by MoOx** [comment]
No worry, will do later.

**Comment by MoOx** [comment]
I found the issue. Currently we [pass imported layouts from the memory](https://github.com/MoOx/statinamic/blob/3fabb759363fa536ac518a1b95eadb291733ebef/src/builder/server.js#L159) and we do not require/import them on rendering time. That's why emptying the cache is useless.
We need to find a way to "refresh" layouts memory (layouts or other things).

**Comment by MoOx** [comment]
The things we should reload are loaded in scripts/build.js which is the file that start the server. We cannot reload the pieces we want with the current architecture.
We must change something and (for example) create a file that export those things (layouts/meta/routes/store) and then import this file via require() just right before the server rendering. This way we can clean node cache before the require() and so get a fresh in memory version.
I think this require some work for not a lot of benefit so I won't handle that asap.

**Comment by thangngoc89** [comment]
Great to hear. I think we can just `require.resolve` layouts instead of import it. And in url-to-html we will check if this is a string,  import it.

And you're right. This should not be high priority.

**Comment by thangngoc89** [comment]
Also,  i think it's time to think about architecture one more time. You know for hot reload markdown.

**Comment by thangngoc89** [comment]
Alright, so I made it work with a minimal change

https://github.com/MoOx/statinamic/compare/master...thangngoc89:clean-node-cache?expand=1

1 drawback is after change a file, full browser refresh take 6s (too slow imho). But I don't think it is important since we have hmr on client side

**Comment by MoOx** [comment]
If we clean the cache, it will be slow. No choice. But at least it's not buggy, and like your said, hmr take the relay when you develop. So no big deal imo.

Instead of 3 filepaths, maybe we should introduce an "web_modules/app/exports.js" that will export all the 3/4 things we are passing for both client and server rendering. This will avoid us to send those 3/4 fields everywhere... Like you said anyway architecture will probably be changed a little bit for markdown hot loading/internal linking.

**Comment by thangngoc89** [comment]
Yet another breaking change :) . I'll send a PR

**Comment by thangngoc89** [comment]
Fixed in #145
